_posts/2019-01-26-generic-way-of-handling-loading-saving-and-validationerrors-in-angular.md

```diff
@@ -0,0 +1,410 @@
+
+```typescript
+loading: boolean;
+acting: boolean;
+
+ngOnInit(): void {
+    this.loading = true;
+    const usersCompleted = false;
+    const citiesCompleted = false;
+    this.userService.fetch().subscribe(users => {
+        usersCompleted = true;
+        this.loading = usersCompleted && citiesCompleted;
+    });
+    this.citiesService.fetch().subscribe(cities => {
+        citiesCompleted = true;
+        this.loading = usersCompleted && citiesCompleted;
+    });
+}
+```
+We have to keep track of which call completes first because we can't set loading to false, if there is call still busy. What happens when there is an error? We would have to implement that as well.
+
+Now this code is only for fetching two lists of data, this becomes ugly pretty quickly and the worst thing about this is that we have to reimplement that for every component that does data fetching.
+
+It becomes even worse if we want to update, add and remove data. Imagine that we have to handle validation errors as well:
+
+```typescript
+remove(user): void {
+    // TODO: set acting to true
+    this.userService.remove(user).subscribe(res => {
+        // TODO: set acting to false
+    })
+}
+update(user): void {
+    // TODO: set acting to true
+    this.userService.update(user).subscribe(res => {
+        // TODO: set acting to false
+        // TODO: handle validation errors
+    })
+}
+add(user): void {
+    // TODO: set acting to true
+    this.userService.add(user).subscribe(res => {
+        // TODO: set acting to false
+        // TODO: handle validation errors
+    })
+}
+```
+
+To handle validation errors we have to check if the httpstatusCode is 400, manually map the data etc.
+
+These codesamples are in this article to prove a point. **It's dirty redundant logic that we have to implement over and over again**
+
+## Let's clean this up
+
+To achieve this we will use an angular *service* in combination with an angular *interceptor* and typescript *decorators*.
+
+The first thing we need is a `HttpStatusService` that exposes 3 observables:
+
+- `loading$`: whether the user is fetching data
+- `acting$`: whether the user is removing, updating or adding data
+- `validationErrors$`: whether there are validation errors or not
+
+```typescript
+// http-status.service.ts
+@Injectable({
+  // important to provide this service to the 
+  // injector of the root module 
+  providedIn: 'root'
+})
+export class HttpStatusService {
+  // regular subject because we don't want to replay
+  // the validationerrors
+  private validationErrorsSub$ 
+    = new Subject<ValidationError[]>();
+
+  // 2 subjects that replays the last value 
+  // (ideal for state)
+  private loadingSub$ = new ReplaySubject<boolean>(1);
+  private actingSub$ = new ReplaySubject<boolean>(1);
+
+  // we don't want to expose the subject for
+  // encapsulation purposes. That's why we convert them
+  // into observables
+  validationErrors$ = 
```

Comment:
Shouldn't work something like this?
```ts
get validationErrors$(): Subject<ValidationError[]> {
   return this.validationErrorsSub$;
}
```

validationErrors$ is readonly in this case

Reply:
ValidationErrors$ is readonly in this case but it still exposes a subject with a next function, that's not what we want. I like the idea of a getter though but I would still use AsObservable